### ISSUE TITLE
implement tools.build.cross_building:cross_build in autotools and cmake toolchains

### ISSUE THREAD

**Issue by jcar87** [conan-io/conan#15969]
TL;DR:  add ability to set this to true (or false) regardless of what Conan might decide, for the super advanced cases. See below.

For context: 
for https://github.com/conan-io/conan/issues/14962 and https://github.com/conan-io/conan/issues/14947

From the documentation:
https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.72/html_node/Specifying-Target-Triplets.html#Specifying-Names

* When `os` and `arch` are both the same - Conan does not pass neither `--host` nor `--build` in Autotools toolchain 
* When cross building (`os`/`arch` are different for build and host), Conan will autoguess some values for _both_ `--host` and `--build`, but these are based on the most "default" cases and may not cover specifics. I can't think of a specific  where one may only one to override the guessed `build` triple - so long as it differs from the host one, autotools will be in cross-compilation mode anyway - but it's good to give the option in case Conan's guess is not the most accurate.
* There can be cases where we have "incompatible" Linux flavour were `os` and `arch` are the same. This could be modelled by a separate setting (e.g. a  subsetting in `os`) or not modelled at all (not recommended!). 
 
@franramirez688, for the third case -  I think there may be a case for replicating what we did in https://github.com/conan-io/conan/pull/15616/files here - in case anyone wants to "force" autotools into cross-building mode (this is telling autotools: do not try to run the executables you are building for the host platform) - that is, force Conan to pass `--build` and `--host` even when `os` and `arch` are the same. It would be around this line https://github.com/ErniGH/conan/blob/35b9792a257c5b4701038d6600ea9fb66d39bc4d/conan/tools/gnu/autotoolstoolchain.py#L71

Arguably the same goes for CMake - CMake assumes cross-compilation (or at least, `CMAKE_CROSSCOMPILING` evaluates to true) when `CMAKE_SYSTEM_NAME` is manually defined. We already have this as a _conf_, so technically it's already possible without touching recipes - but  we may want to listen to `tools.build.cross_building:cross_build` and "force" it as well (I saw a case recently in an issue :D ) Just a thought!

_Originally posted by @jcar87 in https://github.com/conan-io/conan/issues/15965#issuecomment-2023326149_

**Comment by jasal82** [comment]
Would it be possible to get this fix into the next milestone? And a backport to 1.x would also be really appreciated.

**Comment by memsharded** [comment]
Hi @jasal82 

This hasn't been moved forward because it is still not very clear.

For example for ``CMake``, setting ``tools.build.cross_building:cross_build=True`` what would be the effect that is not happening right now? Because CMake as @jcar87 comments above is using the ``CMAKE_SYSTEM_NAME``, but this variable is defined with 

```
system_name = self._conanfile.conf.get("tools.cmake.cmaketoolchain:system_name")
system_version = self._conanfile.conf.get("tools.cmake.cmaketoolchain:system_version")
system_processor = self._conanfile.conf.get("tools.cmake.cmaketoolchain:system_processor")
```

So setting ``tools.build.cross_building:cross_build=True`` doesn't have any effect.

For other build systems, e.g. ``AutotoolsToolchain`` and ``MesonToolchain`` it seems that it has been already done, so it would be necessary to have an updated report for every individual build system, clarifying what would be the expected outcome of setting it.

**Comment by jasal82** [comment]
There are packages in Conan Center (e.g. Google `benchmark`) which fail to build without `cross_build=True` in some cases. For example, we're using a toolchain which cross-compiles from x86_64 to x86_64, but the resulting executables cannot be run on the build machine (because of difference in ld-linux naming and location). The `benchmark` regex feature tests in CMake try to compile **and run** test programs and inspect their exit code. This is fixed for cross compilation toolchains by disabling the feature tests from the conanfile but this is dependent on the return value of `cross_building()`.

**Comment by jasal82** [comment]
And maybe I've misunderstood this issue. I'd need a conf setting that changes the value of `cross_building()` in the conanfile.

**Comment by memsharded** [comment]
This is the current implementation of ``cross_building()``:
```python
def cross_building(conanfile=None, skip_x64_x86=False):
    cross_build = conanfile.conf.get("tools.build.cross_building:cross_build", check_type=bool)
    if cross_build is not None:
        return cross_build
```
If the conf is defined, it will return the conf, so it seems this is already doing what you expect.

**Comment by jasal82** [comment]
Ah great, is that a new feature? Seems I totally missed that.

**Comment by jasal82** [comment]
Unfortunately it's not on 1.x. Can we have a backport please?

**Comment by jasal82** [comment]
#16728

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/16728 merged, thanks for your contribution!
